File: app/src/main/java/com/digis/android/task/data/source/remote/RemoteResponseHandler.java
package com.digis.android.task.data.source.remote;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;

public class RemoteResponseHandler<T> implements RemoteCallback<T> {
    private final Application mApp;
    private final MutableLiveData<Response<T>> mLiveData;

    public RemoteResponseHandler(@NonNull Application app, @NonNull MutableLiveData<Response<T>> liveData) {
        mApp = app;
        mLiveData = liveData;
    }

    @Override
    public final void onResponse(@NonNull HttpResponse<T> httpResponse) {
        onHandleResponse(httpResponse);
    }

    @Override
    public void onFailure(Throwable throwable) {
        mLiveData.setValue(Response.create(mApp, throwable, null));
    }

    @SuppressWarnings("unused")
    protected final MutableLiveData<Response<T>> getLiveData() {
        return mLiveData;
    }

    protected void onHandleResponse(@NonNull HttpResponse<T> httpResponse) {
        mLiveData.setValue(Response.create(httpResponse));
    }
}

File: app/src/main/java/com/digis/android/task/implementation/kotlin/data/source/remote/RemoteResponseHandler.kt
package com.digis.android.task.implementation.kotlin.data.source.remote

import android.app.Application
import androidx.lifecycle.MutableLiveData

open class RemoteResponseHandler<T>(
        private val app: Application,
        private val liveData: MutableLiveData<Response<T>>
) : RemoteCallback<T> {
    override fun onResponse(httpResponse: HttpResponse<T>) {
        liveData.value = Response.create(httpResponse)
    }

    override fun onFailure(throwable: Throwable) {
        liveData.value = Response.create(app, throwable, null)
    }
}